File: Adapter and Facade/Adapter and Facade/DotNet/Adapter and Facade/Adapter and Facade.cs
using Adapter_and_Facade.Ducks;
using Adapter_and_Facade.Home_Theater;
using Adapter_and_Facade.Turkeys;
using System.Numerics;

namespace Adapter_and_Facade;

public class TurkeyAdapter : IDuck
{
    public ITurkey Turkey;

    public TurkeyAdapter(ITurkey turkey)
    {
        Turkey = turkey;
    }

    public void Quack()
    {
        Turkey.Gobble();
    }

    public void Fly()
    {
        for (int i = 0; i < 5; i++)
        {
            Turkey.Fly();
        }
    }
}

public class DuckAdapter : ITurkey
{
    public IDuck Duck;

    public DuckAdapter(IDuck duck)
    {
        Duck = duck;
    }

    public void Gobble()
    {
        Duck.Quack();
    }

    public void Fly()
    {
        Duck.Fly();
    }
}

public class HomeTheaterFacade
{
    public Amplifier Amplifier;
    public Tuner Tuner;
    public StreamingPlayer Player;
    public Projector Projector;
    public TheaterLights Lights;
    public Screen Screen;
    public PopcornPopper Popper;

    public HomeTheaterFacade(
        Amplifier amplifier,
        Tuner tuner,
        StreamingPlayer player,
        Projector proyector,
        Screen screen,
        TheaterLights lights,
        PopcornPopper popper
    )
    {
        Amplifier = amplifier;
        Tuner = tuner;
        Player = player;
        Projector = proyector;
        Lights = lights;
        Screen = screen;
        Popper = popper;
    }

    public void WatchMovie(string movie)
    {
        Console.WriteLine("Get ready to watch a movie...");
        Popper.On();
        Popper.Pop();
        Lights.Dim(10);
        Screen.Down();
        Projector.On();
        Projector.WideScreenMode();
        Amplifier.On();
        Amplifier.Player = Player;
        Amplifier.SetSurroundSound();
        Amplifier.Volume = 5;
        Player.On();
        Player.Play(movie);
    }

    public void EndMovie()
    {
        Console.WriteLine("Shutting movie theater down...");
        Popper.Off();
        Lights.On();
        Screen.Up();
        Projector.Off();
        Amplifier.Off();
        Player.Stop();
        Player.Off();
    }
}

File: Adapter and Facade/Adapter and Facade/Python/src/AdapterAndFacade.py
import random

from .HomeTheater import (
    Tuner,
    Amplifier,
    PopcornPopper,
    Screen,
    StreamingPlayer,
    TheaterLights,
    Projector
)
from .Duck import Duck
from .Turkey import Turkey


class TurkeyAdapter:
 
    def __init__(self, turkey: Turkey) -> None:
        self.turkey = turkey

    def quack(self) -> None:
        self.turkey.gobble()

    def fly(self) -> None:
        for i in range(5):
            self.turkey.fly()


class DuckAdapter:

    def __init__(self, duck: Duck) -> None:
        self.duck = duck

    def gobble(self) -> None:
        self.duck.quack()

    def fly(self) -> None:
        if random.randint(0, 4) == 4:
            self.duck.fly()


class HomeTheaterFacade:
    def __init__(
            self,
            amplifier: Amplifier,
            tuner: Tuner,
            player: StreamingPlayer,
            projector: Projector,
            screen: Screen,
            lights: TheaterLights,
            popper: PopcornPopper
        ):
            self.amplifier = amplifier
            self.tuner = tuner
            self.player = player
            self.projector = projector
            self.screen = screen
            self.lights = lights
            self.popper = popper

    def watch_movie(self, movie: str) -> None:
        print("Get ready to watch a movie...")
        self.popper.on()
        self.popper.pop()
        self.lights.dim(10)
        self.screen.down()
        self.projector.on()
        self.projector.wide_screen_mode()
        self.amplifier.on()
        self.amplifier.player = self.player
        self.amplifier.set_surround_sound()
        self.amplifier.volume = 5
        self.player.on()
        self.player.play(movie)

    def end_movie(self) -> None:
        print("Shutting movie theater down...")
        self.popper.off()
        self.lights.on()
        self.screen.up()
        self.projector.off()
        self.amplifier.off()
        self.player.stop()
        self.player.off()